### PR TITLE
Transaction signing with secrets provided (/wallet/transaction/sign)

### PR DESCRIPTION
**Pull request by kushti**
This PR is introducing:
 * A hierarchy for secret keys, starting from basic SecretKey trait. The hierarchy combines primitive (DLOG/DHT) and HD(BIP32) keys  
 * ErgoProvingInterpreter modified to be friendly to the hierarchy above
 * New Wallet API method /wallet/transaction/signWithSecretsGiven to sign an arbitrary transaction with secrets provided (both DLOG and DHT)

Based on #1085 , related #901 and #940.

### CODE REVIEW

**Comment by jasondavies** [comment]
> New Wallet API method /wallet/transaction/signWithSecretsGiven to sign an arbitrary transaction with secrets provided (both DLOG and DHT)

I feel like `/wallet/transaction/sign` would be a slightly simpler and more general name.  See my next point:

Does this support signing a transaction using the wallet's secrets (i.e. without providing extra secrets in the signing request)?  I think the API should allow signing an arbitrary transaction using the wallet's secrets and optional extra secrets sent via the API.  I tried signing a transaction just now without providing any secrets and it didn't seem to work.

File: src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.ergoplatform.wallet.secrets.{DhtSecretWrapper, DlogSecretWrapper, PrimitiveSecretKey}
+
+/**
+  * Basic trait for externally provided hint for interpreter (to be used during script execution and signature
+  * generation).
+  */
+trait Hint
+
+/**
+  * Externally provided secret (to be used once for a transaction to sign)
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala:15]: Consider clarifying what "used once" means. Does it mean the secret cannot be used to sign two transactions. How it is implemented?

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoProvingInterpreter.scala [hunk 2]
```diff
@@ -5,31 +5,40 @@ import java.util
...
+    * Interpreter's secrets, in form of sigma protocols private inputs
+    */
+  val secrets: IndexedSeq[SigmaProtocolPrivateInput[_, _]] = secretKeys.map(_.key)
 
-  val pubKeys: IndexedSeq[ProveDlog] = secrets.map(_.publicImage)
+  /**
+    * Only secrets corresponding to hierarchical deterministic scheme (BIP-32 impl)
+    */
+  val hdKeys: IndexedSeq[ExtendedSecretKey] =
+    secretKeys.filter(_.isInstanceOf[ExtendedSecretKey]).map(_.asInstanceOf[ExtendedSecretKey])
```
[the remark this answers is not in the document]
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoProvingInterpreter.scala:36]: done

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/SecretKey.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
+package org.ergoplatform.wallet.secrets
+
+import sigmastate.basics.DLogProtocol.DLogProverInput
+import sigmastate.basics.{DiffieHellmanTupleProverInput, SigmaProtocolPrivateInput}
+
+/**
+  * Basic trait for secret data, encapsulating a corresponding private inputs for a Sigma protocol.
+  */
+trait SecretKey {
+  def key: SigmaProtocolPrivateInput[_, _]
```
[the remark this answers is not in the document]
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/SecretKey.scala:10]: done

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/SecretKey.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  * @param key - secret (in form of a sigma-protocol private input)
+  */
+case class DlogSecretWrapper(override val key: DLogProverInput) extends PrimitiveSecretKey
+
+/**
+  * Secret exponent of a Diffie-Hellman tuple, i.e. secret w such as u = g^^w and v = h^^w, where g and h are group
+  * generators, (g,h,u,v) is a public input (public key).
+  * @param key - secret (in form of a sigma-protocol private input)
+  */
+case class DhtSecretWrapper(override val key: DiffieHellmanTupleProverInput) extends PrimitiveSecretKey
```
[the remark this answers is not in the document]
Comment (kushti) [reply ergo-wallet/src/main/scala/org/ergoplatform/wallet/secrets/SecretKey.scala:36]: done

File: src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala (hunk 1 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala:15]: done

**Comment by kushti** [comment]
@aslesarenko all the comments were addressed, please make another pass!

File: src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  lazy val dlogs: Seq[DlogSecretKey] = hints.flatMap{ h => h match{
+    case ExternalSecret(d: DlogSecretKey) => Some(d)
+    case _ => None
+  }}
+
+  lazy val dhts: Seq[DhtSecretKey] = hints.flatMap{ h => h match{
+    case ExternalSecret(d: DhtSecretKey) => Some(d)
+    case _ => None
+  }}
```
Comment (oskin1) [inline src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala:40]:
```suggestion
  lazy val dhts: Seq[DhtSecretKey] = hints.flatMap {
    case ExternalSecret(d: DhtSecretKey) => Some(d)
    case _ => None
  }
```

File: src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+/**
+  * A request to sign a transaction
+  * @param utx - unsigned transaction
+  * @param hints - hints for interpreter (such as additional one-time secrets)
+  * @param inputs - hex-encoded input boxes bytes for the unsigned transaction (optional)
+  * @param dataInputs - hex-encoded data-input boxes bytes for the unsigned transaction (optional)
+  */
+case class TransactionSigningRequest(utx: UnsignedErgoTransaction,
+                                     hints: Seq[Hint],
+                                     inputs: Option[Seq[String]],
```
Comment (oskin1) [inline src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala:29]: Can't the same semantics be encoded by just `Seq[String]`?

Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala:31]: @oskin1 @aslesarenko semantics of Option[Seq[...]] differs from Seq[...], and this is taken in account by the JSON (de)serializer. No need to comment here, denoting optional values via types is standard in Scala.

File: src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala (hunk 5 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala]: simplified even more via collect()

**Comment by kushti** [comment]
@jasondavies wallet and provided secrets now combined in a single /sign method
@oskin1 @aslesarenko comments addressed, please make another pass!

File: src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala (hunk 6 above)
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala:31]: @kushti what is the meaning of Some(Seq()), how it is different from None?

Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/requests/TransactionSigningRequest.scala:31]: @aslesarenko Some(Seq()) denotes existence of the empty sequence, None denotes absence of any collection in the value

**Pull request merged**